crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,165 @@
+#[ruma_event(type = "m.secret.request", kind = ToDevice)]
+pub struct SecretRequestEventContent {
+    /// The action for the request, one of `["request", "request_cancellation"]`.
+    ///
+    /// If the action is "request", the name of the secret must also be provided.
+    #[serde(flatten)]
+    pub action: RequestAction,
+
+    /// The ID of the device requesting the event.
+    pub requesting_device_id: DeviceIdBox,
+
+    /// A random string uniquely identifying (with respect to the requester and the target) the
+    /// target for a secret.
+    ///
+    /// If the secret is requested from multiple devices at the same time, the same ID may be used
+    /// for every target. The same ID is also used in order to cancel a previous request.
+    pub request_id: String,
+}
+
+impl SecretRequestEventContent {
+    /// Creates a new `SecretRequestEventContent` with the given action, requesting device ID and
+    /// request ID.
+    pub fn new(
+        action: RequestAction,
+        requesting_device_id: DeviceIdBox,
+        request_id: String,
+    ) -> Self {
+        Self { action, requesting_device_id, request_id }
+    }
+}
+
+/// Action for a *m.secret.request* event.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum RequestAction {
+    /// Request a secret by its name.
+    Request(SecretName),
+
+    /// Cancel a request for a secret.
+    RequestCancellation,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+impl Serialize for RequestAction {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Request(name) => {
+                let mut st = serializer.serialize_map(Some(2))?;
+                st.serialize_entry("name", name)?;
+                st.serialize_entry("action", "request")?;
+                st.end()
+            }
+            Self::RequestCancellation => {
+                let mut st = serializer.serialize_map(Some(1))?;
+                st.serialize_entry("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => serializer.serialize_str(custom),
+        }
+    }
+}
+
+/// The name of a secret.
+#[derive(Clone, Debug, StringEnum)]
+pub enum SecretName {
+    /// Cross-signing master key (m.cross_signing.master).
+    #[ruma_enum(rename = "m.cross_signing.master")]
+    CrossSigningMasterKey,
+
+    /// Cross-signing user-signing key (m.cross_signing.user_signing).
+    #[ruma_enum(rename = "m.cross_signing.user_signing")]
+    CrossSigningUserSigningKey,
+
+    /// Cross-signing self-signing key (m.cross_signing.self_signing).
+    #[ruma_enum(rename = "m.cross_signing.self_signing")]
+    CrossSigningSelfSigningKey,
+
+    /// Recovery key (m.megolm_backup.v1.recovery_key).
+    #[ruma_enum(rename = "m.megolm_backup.v1.recovery_key")]
```

Comment:
Is this right? I can't find a mention of `m.megolm_backup.v1.recovery_key` in the spec, as far as I can tell the name is just `m.megolm_backup.v1` mentioned over [here](https://spec.matrix.org/unstable/client-server-api/#recovery-key):

> The recovery key can also be stored on the server or shared with other devices using the Secrets module. When doing so, it is identified using the name m.megolm_backup.v1, and the key is base64-encoded before being encrypted.

Reply:
For the key names I had only based myself off of the MSC text [here](https://github.com/uhoreg/matrix-doc/blob/ssss/proposals/1946-secure_server-side_storage.md#secret-storage):

> Encrypted data is stored in the user's account_data using the event type defined by the feature that uses the data. For example, decryption keys for key backups could be stored under the type `m.megolm_backup.v1.recovery_key`, or the self-signing key for cross-signing could be stored under the type `m.cross_signing.self_signing`.

It does seem that it should be `m.megolm_backup.v1`, since that's what's used in the more recent MSCs, the unstable spec and matrix-js-sdk.

Reply:
Right, let's go with what's in the unstable spec and in the js-sdk.